File: mobile/src/services/quotation.ts
import { apiClient } from '@/lib/api';
import { Quotation, QuotationItem, QuotationStatus } from '@/types';

export interface CreateQuotationItemInput {
  description: string;
  quantity: number;
  unit_price: number;
  tax_rate: number;
}

export interface CreateQuotationInput {
  customer_id: string;
  issue_date: string;
  valid_until: string;
  discount?: number;
  notes?: string;
  terms?: string;
  items: CreateQuotationItemInput[];
}

interface ApiQuotationItem {
  id: string;
  quotation_id: string;
  description: string;
  quantity: string | number;
  unit_price: string | number;
  tax_rate: string | number;
  tax_amount: string | number;
  line_total: string | number;
}

interface ApiQuotation {
  id: string;
  business_id: string;
  customer_id: string;
  customer_name?: string | null;
  quotation_number: string;
  issue_date: string;
  valid_until: string;
  status: QuotationStatus;
  subtotal: string | number;
  discount: string | number;
  tax: string | number;
  total: string | number;
  notes?: string | null;
  terms?: string | null;
  items: ApiQuotationItem[];
  created_at: string;
  updated_at: string;
}

function mapApiItem(item: ApiQuotationItem): QuotationItem {
  return {
    id: item.id,
    quotationId: item.quotation_id,
    description: item.description,
    quantity: Number(item.quantity),
    unitPrice: Number(item.unit_price),
    taxRate: Number(item.tax_rate),
    taxAmount: Number(item.tax_amount),
    lineTotal: Number(item.line_total),
  };
}

function mapApiQuotation(q: ApiQuotation): Quotation {
  return {
    id: q.id,
    businessId: q.business_id,
    customerId: q.customer_id,
    customerName: q.customer_name || 'Client',
    quotationNumber: q.quotation_number,
    issueDate: q.issue_date,
    validUntil: q.valid_until,
    status: q.status,
    subtotal: Number(q.subtotal),
    discount: Number(q.discount),
    tax: Number(q.tax),
    total: Number(q.total),
    notes: q.notes ?? undefined,
    terms: q.terms ?? undefined,
    items: (q.items || []).map(mapApiItem),
  };
}

export const QuotationService = {
  /**
   * Fetch all quotations for the current business.
   */
  async getQuotations(status?: QuotationStatus): Promise<Quotation[]> {
    const params = status && status !== ('ALL' as unknown) ? { status } : {};
    const response = await apiClient.get<ApiQuotation[]>('/api/v1/quotations', { params });
    return response.data.map(mapApiQuotation);
  },

  /**
   * Fetch single quotation by ID.
   */
  async getQuotationById(id: string): Promise<Quotation> {
    const response = await apiClient.get<ApiQuotation>(`/api/v1/quotations/${id}`);
    return mapApiQuotation(response.data);
  },

  /**
   * Create a new quotation with server-authoritative calculations.
   */
  async createQuotation(input: CreateQuotationInput): Promise<Quotation> {
    const payload = {
      customer_id: input.customer_id,
      issue_date: input.issue_date,
      valid_until: input.valid_until,
      discount: (input.discount || 0).toFixed(2),
      notes: input.notes,
      terms: input.terms,
      items: input.items.map((i) => ({
        description: i.description,
        quantity: i.quantity.toFixed(2),
        unit_price: i.unit_price.toFixed(2),
        tax_rate: i.tax_rate.toFixed(2),
      })),
    };
    const response = await apiClient.post<ApiQuotation>('/api/v1/quotations', payload);
    return mapApiQuotation(response.data);
  },
};

File: backend/app/models/quotation.py
import enum
import uuid
from datetime import date
from decimal import Decimal
from typing import Optional, TYPE_CHECKING
from sqlalchemy import String, Text, Date, Numeric, Enum as SQLEnum, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.base import UUIDModel

if TYPE_CHECKING:
    from app.models.business import Business
    from app.models.customer import Customer


class QuotationStatus(str, enum.Enum):
    DRAFT = "DRAFT"
    SENT = "SENT"
    ACCEPTED = "ACCEPTED"
    REJECTED = "REJECTED"
    EXPIRED = "EXPIRED"
    CONVERTED = "CONVERTED"


class Quotation(UUIDModel):
    """
    Commercial quotation entity.
    Enforces Rule 4 (Financial Exact Decimals): All currency and subtotal columns
    use NUMERIC(12, 2) rather than floating point numbers.
    """

    __tablename__ = "quotations"

    business_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("businesses.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )

    customer_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("customers.id", ondelete="RESTRICT"),
        index=True,
        nullable=False,
    )

    quotation_number: Mapped[str] = mapped_column(
        String(50),
        index=True,
        nullable=False,
    )

    issue_date: Mapped[date] = mapped_column(
        Date,
        nullable=False,
    )

    valid_until: Mapped[date] = mapped_column(
        Date,
        nullable=False,
    )

    status: Mapped[QuotationStatus] = mapped_column(
        SQLEnum(QuotationStatus, name="quotation_status_enum"),
        default=QuotationStatus.DRAFT,
        index=True,
        nullable=False,
    )

    # Financial figures - Strict NUMERIC(12, 2)
    subtotal: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    discount: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    tax: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    total: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    notes: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )

    terms: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )

    # Relationships
    business: Mapped["Business"] = relationship("Business", back_populates="quotations")
    customer: Mapped["Customer"] = relationship("Customer", back_populates="quotations")
    items: Mapped[list["QuotationItem"]] = relationship(
        "QuotationItem",
        back_populates="quotation",
        cascade="all, delete-orphan",
        order_by="QuotationItem.created_at.asc()",
    )


class QuotationItem(UUIDModel):
    """
    Individual line item for a quotation.
    Quantities, unit prices, tax rates, tax amounts, and line totals all use NUMERIC(12, 2).
    """

    __tablename__ = "quotation_items"

    quotation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("quotations.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )

    description: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
    )

    quantity: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("1.00"),
        nullable=False,
    )

    unit_price: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    tax_rate: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    tax_amount: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    line_total: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        default=Decimal("0.00"),
        nullable=False,
    )

    # Relationships
    quotation: Mapped["Quotation"] = relationship("Quotation", back_populates="items")
